plugins/org.python.pydev/src/org/python/pydev/editor/correctionassist/heuristics/AssistFString.java

```diff
@@ -0,0 +1,105 @@
+package org.python.pydev.editor.correctionassist.heuristics;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.eclipse.jface.text.BadLocationException;
+import org.eclipse.jface.text.IDocument;
+import org.eclipse.jface.text.ITypedRegion;
+import org.python.pydev.core.IPyEdit;
+import org.python.pydev.core.IPythonNature;
+import org.python.pydev.core.MisconfigurationException;
+import org.python.pydev.core.docutils.PySelection;
+import org.python.pydev.core.partition.PyPartitionScanner;
+import org.python.pydev.core.proposals.CompletionProposalFactory;
+import org.python.pydev.editor.PyEdit;
+import org.python.pydev.editor.codefolding.PySourceViewer;
+import org.python.pydev.editor.correctionassist.IAssistProps;
+import org.python.pydev.shared_core.code_completion.ICompletionProposalHandle;
+import org.python.pydev.shared_core.code_completion.IPyCompletionProposal;
+import org.python.pydev.shared_core.image.IImageCache;
+import org.python.pydev.shared_core.image.IImageHandle;
+import org.python.pydev.shared_core.image.UIConstants;
+import org.python.pydev.shared_core.partitioner.FastPartitioner;
+import org.python.pydev.shared_core.string.FastStringBuffer;
+
+public class AssistFString implements IAssistProps {
+    @Override
+    public List<ICompletionProposalHandle> getProps(PySelection ps, IImageCache imageCache, File f,
+            IPythonNature nature, IPyEdit edit, int offset) throws BadLocationException, MisconfigurationException {
+        PySourceViewer viewer = null;
+        if (edit instanceof PyEdit) { //only in tests it is actually null
+            viewer = ((PyEdit) edit).getPySourceViewer();
+        }
+        List<ICompletionProposalHandle> l = new ArrayList<ICompletionProposalHandle>();
+
+        IDocument doc = ps.getDoc();
+        FastPartitioner fastPartitioner = (FastPartitioner) PyPartitionScanner.checkPartitionScanner(doc);
+        ITypedRegion partition = fastPartitioner.getPartition(offset);
+
+        FastStringBuffer strBuf = new FastStringBuffer();
+        strBuf.append('f').append(doc.get(partition.getOffset(), partition.getLength()));
+
+        FastStringBuffer variablesBuf = new FastStringBuffer();
+        // get string end offset
+        int partitionEndOffset = partition.getOffset() + partition.getLength();
+        // get format (% ....) end offset
+        int formatEndOffset = doc.getLineLength(doc.getLineOfOffset(partitionEndOffset));
```

Comment:
I believe this will fail with multi-lines in the variables part.

So, a good test for this is:

```
`"%s - %s" % (
  a,
  b
)
```

The result should be: `f"{a} - {b}"`.

I think you should get the offset end and then get the next non-whitespace and then if it's a name collect the name or if it's a parenthesis collect it through parsingUtils.

Something as the code below (note that I didn't test it, but it should be a good reference).

```
            ParsingUtils parsingUtils = ParsingUtils.create(doc);
            int i = strEndOffset;
            FastStringBuffer bufWithVars = new FastStringBuffer();
            while(i < parsingUtils.len()) {
                char c = parsingUtils.charAt(i);
                if(Character.isWhitespace(c)) {
                    i++;
                    continue;
                }
                if(c == '{' || c == '[' || c == '(' || c == '"' || c == '\'') {
                    i = parsingUtils.eatPar(i, bufWithVars, c);
                    break;
                }
                if(Character.isJavaIdentifierPart(c)) {
                    bufWithVars.append(c);
                    while(i < parsingUtils.len()) {
                        c = parsingUtils.charAt(i);
                        if(Character.isJavaIdentifierPart(c)) {
                            bufWithVars.append(c);
                        }else {
                            if(c == '{' || c == '[' || c == '(' || c == '"' || c == '\'') {
                                i = parsingUtils.eatPar(i, bufWithVars, c);
                            }
                            break; // inner break
                        }
                        i++;
                    }
                    break; // outer break
                }
            }
            // at this point i should have the end offset of the variables
```

Reply:
Oh, in the `Character.isJavaIdentifierPart(c)`, you also need to check for `.`.